src/compiler/checker.ts

```diff
@@ -18997,7 +18997,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // (T | U)[K] -> T[K] & U[K] (writing)
         // (T & U)[K] -> T[K] & U[K]
         if (objectType.flags & TypeFlags.Union || objectType.flags & TypeFlags.Intersection && !shouldDeferIndexType(objectType)) {
-            const types = map((objectType as UnionOrIntersectionType).types, t => getSimplifiedType(getIndexedAccessType(t, indexType), writing));
+            const types = map((objectType as UnionOrIntersectionType).types, t => getSimplifiedType(getIndexedAccessType(getSimplifiedType(t, writing), indexType), writing));

```

Comment:
Tt's important to simplify the `t` (a constituent of the `objectType`) here "eagerly". At this stage `getIndexedAccessType` has a chance to return `unknownType` (thus effectively eliminating this constituent from the overall simplified indexed access type on an intersection, since `T & unknown` is `T`).

Without it, the compiler comes back to `getSimplifiedIndexedAccessType` with the non-simplified indexed access. It simplifies the object type there, but it also decides there is nothing to simplify further so it returns the original non-simplified `type`. That leads to a relationship error.